client/client.go

```diff
@@ -448,11 +437,28 @@ func (r *NotaryRepository) GetTargetByName(name string, roles ...string) (*Targe
 	if len(roles) == 0 {
 		roles = append(roles, data.CanonicalTargetsRole)
 	}
+	var resultMeta data.FileMeta
+	var resultRoleName string
 	for _, role := range roles {
-		meta, foundRole := c.TargetMeta(role, name, roles...)
-		if meta != nil {
-			return &TargetWithRole{
-				Target: Target{Name: name, Hashes: meta.Hashes, Length: meta.Length}, Role: foundRole}, nil
+		// Define a visitor function to find the specified target
+		getTargetVisitorFunc := func(tgt *data.SignedTargets, validRole data.DelegationRole) error {
+			var ok bool
+			if tgt == nil {
+				return tuf.ErrContinueWalk{}
+			}
+			// We found the target and validated path compatibility in our walk,
+			// so we should stop our walk and set the resultMeta and resultRoleName variables
+			// Also check that we should not exclude this target (see IMPORTANT comment above)
+			shouldExclude := utils.StrSliceContains(roles, validRole.Name) && validRole.Name != role
+			if resultMeta, ok = tgt.Signed.Targets[name]; ok || shouldExclude {
+				resultRoleName = validRole.Name
+				return tuf.ErrStopWalk{}
+			}
+			return tuf.ErrContinueWalk{}
+		}
+		err = r.tufRepo.WalkTargets(name, role, getTargetVisitorFunc)
```

Comment:
When the error is not nil, we don't need to do anything with it because we intended to walk the entire targets tree, though we should also check if the target we "found" is not nil